### ISSUE TITLE
`test_pthread_proxying_cpp` hangs when used with `-sEXIT_RUNTIME=0`

### ISSUE THREAD

**Issue by donalffons** [emscripten-core/emscripten#18667]
I'm studying emscripten's proxying system to understand how / if it's applicable / helpful to my use case and ran into the following issue.

When trying to execute the test `test_pthread_proxying_cpp`, the execution hangs at [this line](https://github.com/emscripten-core/emscripten/blob/main/test/pthread/test_pthread_proxying_cpp.cpp#L73) forever with `-sEXIT_RUNTIME=0`. I.e. the lambda function is seemingly never executed (I tried both `cout`ing and throwing from the first line of that function in conjunction with `-fwasm-exceptions`). When I use`-sEXIT_RUNTIME=1`, the test completes as expected.

Here is my cmake config:

```cmake
cmake_minimum_required(VERSION 3.5)

project(TestApplication)

add_executable(TestApplication "test_pthread_proxying_cpp.cpp")

set_target_properties(
  TestApplication PROPERTIES
  COMPILE_FLAGS "-pthread"
  LINK_FLAGS "-pthread -sPROXY_TO_PTHREAD=1"
)
```

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.31-git (e88336121cfe6da4a96c88e46f314552f07dfed0)
clang version 17.0.0 (/startdir/llvm-project 1142e6c7c795de7f80774325a07ed49bc95a48c9)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /opt/emscripten-llvm/bin

**Comment by tlively** [comment]
Hmm, the proxied work isn't being executed because it arrives before `pthread_self` is installed...

Investigating further!

**Comment by tlively** [comment]
Ok, I tracked this down to the fact that `emscripten_keepalive_push` is a no-op when EXIT_RUNTIME=0, so the "returner" thread in that test isn't actually staying alive when it returns to its event loop. Since it doesn't stay alive, its pthread_self pointer gets cleared as soon as the thread main function returns, so subsequent proxied work that arrives does not get executed, causing the test to hang.

Assigning to @sbc100 to make `emscripten_keepalive_push` properly keep threads alive regardless of EXIT_RUNTIME.

FWIW, the C version of this test doesn't have the same problem because it uses `emscripten_exit_with_live_runtime` rather than `emscripten_keepalive_push`.

**Comment by sbc100** [comment]
This was fixed by #18785
